packages/utilities/src/classNamesFunction.ts

```diff
@@ -12,29 +25,114 @@ export interface IClassNamesFunctionOptions {
+  // we call the `getStyles` api to evaluate, cache on the property, and return that.
+  let map: IRecursiveMemoNode = new Map();
+  let resultCount = 0;
 
   const getClassNames = (
     styleFunctionOrObject: IStyleFunctionOrObject<TStyleProps, TStyleSet> | undefined,
     styleProps: TStyleProps = {} as TStyleProps
   ): IProcessedStyleSet<TStyleSet> => {
-    // styleSet might be undefined if styleFunctionOrObject is undefined, but getStyles should never
-    // ordinarily be undefined (it would hardly make any sense).
-    // However, because we usually use `props.styles` as the argument to an invocation of this method, and
-    // `props.styles` itself is defined as optional, this avoids the need to use `!` at all invocation points.
-    if (styleFunctionOrObject === undefined) {
-      return {} as IProcessedStyleSet<TStyleSet>;
+    let current: Map<any, any> = map;
+    const disableCaching = options.disableCaching;
+
+    if (!options.disableCaching) {
+      current = _traverseMap(map, styleFunctionOrObject as any);
+      current = _traverseMap(current, styleProps);
     }
 
-    const styleSet =
-      styleFunctionOrObject && (typeof styleFunctionOrObject === 'function' ? styleFunctionOrObject(styleProps!) : styleFunctionOrObject);
+    if (disableCaching || !(current as any)[RetVal]) {
+      if (styleFunctionOrObject === undefined) {
+        (current as any)[RetVal] = {} as IProcessedStyleSet<TStyleSet>;
+      } else if (Array.isArray(styleFunctionOrObject)) {
+        (current as any)[RetVal] = mergeStyleSets(...(styleFunctionOrObject as any).map(_derive, styleProps));
+      } else {
+        (current as any)[RetVal] = mergeStyleSets(
+          typeof styleFunctionOrObject === 'function' ? styleFunctionOrObject(styleProps) : styleFunctionOrObject
+        );
+      }
+
+      if (!disableCaching) {
+        resultCount++;
+      }
+    }
+
+    if (resultCount > MAX_CACHE_COUNT) {
+      map.clear();
+      resultCount = 0;
+
+      // Mutate the options passed in, that's all we can do.
+      options.disableCaching = true;
+
+      // Note: this code is great for debugging problems with styles being recaculated, but commenting it out
+      // to avoid confusing consumers.
+
+      // if (process.env.NODE_ENV !== 'production') {
+      //  console.log('Styles are being recalculated far too frequently. Something is mutating the class over and over.');
+      //  // tslint:disable-next-line:no-console
+      //  console.trace();
+      // }
+    }
 
-    return mergeStyleSets(styleSet as TStyleSet);
+    // Note: the RetVal is an attached property on the Map; not a key in the Map. We use this attached property to
+    // cache the return value for this branch of the graph.
+    return (current as any)[RetVal];
   };
 
   return getClassNames;
 }
+
+function _derive(obj: any): any {
+  return typeof obj === 'function' ? obj(this) : obj;
```

Comment:
`this` still confuses me a bit, in this case, what is it actually referring to? If it's a function then it's supposed to be a styleFunction right? And `this` should refer to style props? Am I reading that right?